### PR TITLE
Allow `.sizeBy()` to filter for any option

### PR DESCRIPTION
**Pull request by slavb18**
queue.sizeBy currently supports filtering by priority only
it could be extended to any prop
e.g.
```javascript
const options = { uid: '341a0c1b-147a-4b8e-9b9d-186773918a25' };
queue.add(() => delay(2000), options);
console.log('size by', queue.sizeBy({ uid: '341a0c1b-147a-4b8e-9b9d-186773918a25' }));
```

### CODE REVIEW

File: source/shallow-equal.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+export function shallowPartialEqual(object1: any, object2: any) {
+	const keys1 = Object.keys(object1);
+	for (const key of keys1) {
```
Comment (sindresorhus) [inline source/shallow-equal.ts:3]: Use `Object.entries`. It can also just be inline here.

File: source/shallow-equal.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+export function shallowPartialEqual(object1: any, object2: any) {
```
Comment (sindresorhus) [inline source/shallow-equal.ts:1]:
```suggestion
export function shallowPartialEqual(object1: object, object2: object) {
```

File: source/priority-queue.ts [hunk 3]
```diff
@@ -16,8 +17,8 @@ export default class PriorityQueue implements Queue<RunFunction, PriorityQueueOp
 		};
 
 		const element = {
-			priority: options.priority,
 			run,
+			...options,
```
Comment (sindresorhus) [inline source/priority-queue.ts:21]: I don't think it should be able to overwrite `run`.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
Can you add a test and docs updates?

File: test/test.ts [hunk 4]
```diff
@@ -118,6 +118,19 @@ test('.sizeBy() - priority', async t => {
...
+	queue.add(async () => 0, {uid: '07227c1e-0476-46b5-bd5b-5436626ede16'});
+	queue.add(async () => 0, {uid: '341a0c1b-147a-4b8e-9b9d-186773918a25'});
+	t.is(queue.sizeBy({uid: '341a0c1b-147a-4b8e-9b9d-186773918a25'}), 2);
+	t.is(queue.sizeBy({uid: '07227c1e-0476-46b5-bd5b-5436626ede16'}), 1);
+	queue.clear();
+	await queue.onEmpty();
+	t.is(queue.sizeBy({uid: '341a0c1b-147a-4b8e-9b9d-186773918a25'}), 0);
+	t.is(queue.sizeBy({uid: '07227c1e-0476-46b5-bd5b-5436626ede16'}), 0);
+});
 test('.add() - timeout without throwing', async t => {
```
Comment (sindresorhus) [inline test/test.ts:134]:
```suggestion
});

test('.add() - timeout without throwing', async t => {
```

File: readme.md [hunk 5]
```diff
@@ -131,6 +131,12 @@ Default: `0`
 
 Priority of operation. Operations with greater priority will be scheduled first.
 
+##### any other key
+
+Type: `object`
+
+You can pass any keys in options to use then later in sizeBy filtering
```
Comment (sindresorhus) [inline readme.md:138]: There should rather be a description under `#### options`, that you can pass any other key.

Comment (sindresorhus) [inline readme.md:138]:
```suggestion
##### any other key

Type: `object`

You can pass any keys in options to use then later in `sizeBy` filtering.
```

Also this sentence doesn't read well.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
Bump :)

**Comment by sindresorhus** [comment]
Tests are not passing.

File: readme.md [hunk 6]
```diff
@@ -124,6 +124,7 @@ Promise-returning/async function. When executed, it will receive `{signal}` as t
 
 Type: `object`
 
+You can pass any keys in options to use then later in `sizeBy` filtering.
```
Comment (Richienb) [inline readme.md:127]: There should be an extra newline after this

File: source/shallow-equal.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+export function shallowPartialEqual(object1: Record<string, unknown>, object2: Record<string, unknown>) {
+	for (const [key, value] of Object.entries(object1)) {
```
Comment (Richienb) [inline source/shallow-equal.ts:2]: This could be done in one line with `.every()`

File: test/test.ts [hunk 8]
```diff
@@ -118,6 +118,20 @@ test('.sizeBy() - priority', async t => {
...
+	queue.pause();
+	queue.add(async () => 0, {uid: '341a0c1b-147a-4b8e-9b9d-186773918a25'});
+	queue.add(async () => 0, {uid: '07227c1e-0476-46b5-bd5b-5436626ede16'});
+	queue.add(async () => 0, {uid: '341a0c1b-147a-4b8e-9b9d-186773918a25'});
+	t.is(queue.sizeBy({uid: '341a0c1b-147a-4b8e-9b9d-186773918a25'}), 2);
+	t.is(queue.sizeBy({uid: '07227c1e-0476-46b5-bd5b-5436626ede16'}), 1);
+	queue.clear();
+	await queue.onEmpty();
+	t.is(queue.sizeBy({uid: '341a0c1b-147a-4b8e-9b9d-186773918a25'}), 0);
+	t.is(queue.sizeBy({uid: '07227c1e-0476-46b5-bd5b-5436626ede16'}), 0);
```
Comment (Richienb) [inline test/test.ts:132]: A simpler id system should be used to make it easier to test. Maybe something like `{id: 'foo'}`.

File: readme.md [hunk 9]
```diff
@@ -214,7 +215,7 @@ Size of the queue, the number of queued items waiting to run.
 
 Size of the queue, filtered by the given options.
 
-For example, this can be used to find the number of items remaining in the queue with a specific priority level.
+For example, this can be used to find the number of items remaining in the queue with a specific priority level or any other field.
```
Comment (Richienb) [inline readme.md:218]: There should be a detailed example given instead of just `any other field`.

**Review by Richienb — CHANGES_REQUESTED** [review]
The types need to be updated

File: readme.md (hunk 6 above)
Comment (Richienb) [reply readme.md:127]: Should be reworded to make it more direct

**Comment by Richienb** [comment]
Bump :)

**Pull request closed without merging**